From 4154e818c885cc87e52519215c552fa8e620ad58 Mon Sep 17 00:00:00 2001
From: Raul E Rangel <rrangel@chromium.org>
Date: Wed, 11 Oct 2023 12:42:34 -0600
Subject: [PATCH] config: Sort USE_EXPAND variables

Without the sort, we end up extending the USE_EXPAND variables with
a non deterministic ordering. This makes binary package creation non
hermetic.

i.e.,
```
-declare -x PYTHON_TARGETS="python3_8 python3_7 python3_6 python3_10 python3_9"
+declare -x PYTHON_TARGETS="python3_8 python3_9 python3_10 python3_7 python3_6"
```

Assuming `PYTHON_TARGETS=python3_8` in make.conf, after this change we
get:
```
declare -x PYTHON_TARGETS="python3_8 python3_10 python3_6 python3_7 python3_9"
```

Ideally we would completely sort the USE_EXPAND variables, but the
LINGUAS appears to need a very specific ordering.

BUG=b:259558106
TEST=build dev-lang/python-exec twice and verify package is identical

Change-Id: I16551c3618ede5c05ba4b5bbe84ad42571ae058a
---
 usr/PYTHON_LIBDIR/PYTHON_VERSION/site-packages/portage/package/ebuild/config.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/usr/PYTHON_LIBDIR/PYTHON_VERSION/site-packages/portage/package/ebuild/config.py b/usr/PYTHON_LIBDIR/PYTHON_VERSION/site-packages/portage/package/ebuild/config.py
index 075a9b1d9d..29155a44b8 100644
--- a/usr/PYTHON_LIBDIR/PYTHON_VERSION/site-packages/portage/package/ebuild/config.py
+++ b/usr/PYTHON_LIBDIR/PYTHON_VERSION/site-packages/portage/package/ebuild/config.py
@@ -1379,7 +1379,7 @@ class config(object):
 			# Preserve the order of var_split because it can matter for things
 			# like LINGUAS.
 			var_split = [ x for x in var_split if x in expand_flags ]
-			var_split.extend(expand_flags.difference(var_split))
+			var_split.extend(sorted(expand_flags.difference(var_split)))
 			has_wildcard = '*' in expand_flags
 			if has_wildcard:
 				var_split = [ x for x in var_split if x != "*" ]
-- 
2.42.0.655.g421f12c284-goog

